pkg/reconciler/garbagecollector/garbagecollector_controller.go

```diff
@@ -0,0 +1,319 @@
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+// processNextWorkItem waits for the queue to have a key available and then processes it.
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	raw, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := raw.(string)
+
+	logger := logging.WithQueueKey(klog.FromContext(ctx), key)
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(1).Info("processing key")
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		utilruntime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", controllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+
+	c.queue.Forget(key)
+
+	return true
+}
+
+// process processes a single key from the queue.
+func (c *Controller) process(ctx context.Context, key string) error {
+	logger := klog.FromContext(ctx)
+	parent, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return nil
+	}
+
+	// turn it into root:org:ws
+	clusterName := parent.Join(name)
+	logger = logger.WithValues("logicalCluster", clusterName.String())
+
+	ws, err := c.getClusterWorkspace(key)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			logger.V(2).Info("ClusterWorkspace not found - stopping garbage collector controller for it (if needed)")
+
+			c.lock.Lock()
+			cancel, ok := c.cancelFuncs[clusterName]
+			if ok {
+				cancel()
+				delete(c.cancelFuncs, clusterName)
+			}
+			c.lock.Unlock()
+
+			c.dynamicDiscoverySharedInformerFactory.Unsubscribe("gc-" + clusterName.String())
+
+			return nil
+		}
+
+		return err
+	}
+	logger = logging.WithObject(logger, ws)
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	_, found := c.cancelFuncs[clusterName]
+	if found {
+		logger.V(4).Info("garbage collector controller already exists")
+		return nil
+	}
+
+	logger.V(2).Info("starting garbage collector controller")
+
+	ctx, cancel := context.WithCancel(ctx)
+	ctx = klog.NewContext(ctx, logger)
+	c.cancelFuncs[clusterName] = cancel
+
+	if err := c.startGarbageCollectorForClusterWorkspace(ctx, clusterName); err != nil {
+		cancel()
+		return fmt.Errorf("error starting garbage collector controller for cluster %q: %w", clusterName, err)
+	}
+
+	return nil
+}
+
+func (c *Controller) startGarbageCollectorForClusterWorkspace(ctx context.Context, clusterName logicalcluster.Name) error {
+	logger := klog.FromContext(ctx)
+
+	kubeClient := c.kubeClusterClient.Cluster(clusterName)
+
+	ignoredResources := make(map[schema.GroupResource]struct{})
+	// Add default ignored resources
+	for gr := range garbagecollector.DefaultIgnoredResources() {
+		ignoredResources[gr] = struct{}{}
+	}
+	// Add projected API resources
+	for gvr := range projection.ProjectedAPIs() {
+		ignoredResources[gvr.GroupResource()] = struct{}{}
+	}
```

Comment:
Could maybe calculate these once and store them as a field on `Controller`? I don't think they change per workspace?

Reply:
Done with 5d46adb24a14aa0b2477f7234472294a3c12db6d.